go-controller/pkg/node/ovn_db_mgmt.go

```diff
@@ -0,0 +1,139 @@
+package node
+
+import (
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+func (n *OvnNode) ensureOvnRaft(db string) {
+	ticker := time.NewTicker(60 * time.Second)
+	klog.Infof("Starting ensure routine for Raft db: %s", db)
+	kclient := n.Kube.(*kube.Kube)
+	for {
+		select {
+		case <-ticker.C:
+			if dbOnNode, _ := metrics.CheckPodRunsOnGivenNode(kclient.KClient.(*kubernetes.Clientset),
+				"name=ovnkube-db", n.name, false); dbOnNode {
+				klog.Infof("Node %s has a OVN DB pod, ensuring OVN Raft membership", n.name)
+				n.ensureLocalRaftServerID(db)
+				n.ensureClusterRaftMembership(db)
+			}
+		case <-n.stopChan:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// ensureLocalRaftServerID is used to ensure there is no stale member in the Raft cluster with our address
+func (n *OvnNode) ensureLocalRaftServerID(db string) {
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+	} else {
+		dbName = "OVN_Southbound"
+		appCtl = util.RunOVNSBAppCtl
+	}
+
+	out, stderr, err := util.RunOVSDBTool("db-sid", db)
+	if err != nil {
+		klog.Warningf("Unable to get db server ID for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	if len(out) < 4 {
+		klog.Errorf("Invalid db id found: %s for db: %s", out, db)
+		return
+	}
+	// server ID in raft membership is only first 4 char prefix
+	serverID := out[:4]
+	out, stderr, err = appCtl("cluster/status", dbName)
+	if err != nil {
+		klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+	matches := r.FindStringSubmatch(out)
+	if len(matches) < 2 {
+		klog.Warningf("Unable to parse Address for db: %s, output: %s", db, out)
+		return
+	}
+	addr := matches[1]
+	// look for current servers in raft cluster with the same address
+	r, _ = regexp.Compile("([a-z0-9]{4}) at " + addr)
+	members := r.FindAllStringSubmatch(out, -1)
+	for _, member := range members {
+		if len(member) < 2 {
+			klog.Warningf("Unable to find server id submatch in %s", member)
+			return
+		}
+		if member[1] != serverID {
+			// stale entry found for this node with same adddress, need to kick
+			klog.Infof("Previous stale member found: %s...kicking", member[1])
+			_, stderr, err = appCtl("cluster/kick", dbName, member[1])
+			if err != nil {
+				klog.Errorf("Error while kicking old Raft member: %s, for address: %s in db: %s,"+
+					"stderr: %v, error: %v", member[1], addr, db, stderr, err)
+			}
```

Comment:
you cannot `kick` yourself out. You will have run `cluster/kick` from the other DB nodes. 

Please take a look at **How to Maintain a Clustered Database** at the link here https://docs.openvswitch.org/en/latest/ref/ovsdb.7/#how-to-maintain-a-clustered-database

Reply:
you can kick yourself out. I tested it

Reply:
That is only if the node has not crashed or down from the cluster or some such. As I understand, `kick` is meant for remaining set of healthy nodes in a cluster to kick an unhealthy node.

Also, the cluster/kick will only work if the cluster is healthy. That is why, we need an observer (aka operator) that looks at the cluster from outside. Figures out what is wrong with the cluster as a whole -- which node is up and which node is down due to corrupted db or crashed db  and so on -- and then makes a informed decision.

Applying fixes in a localized manner will not work.

Reply:
this specific function is for kicking a previous id out of the cluster using the same address as this node's current address. There's no way for an outside operator to know which is the valid one. Only the local node knows which one is his current id.